### PR TITLE
ready for GHC 8.10.2

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [x] ready for GHC 8.10.2

checklist
---------

- [x] compiles (`cabal v2-build` or `stack build`)
- [x] tests run successfully (`cabal v2-test` or `stack test`)
- [ ] documentation added
- [ ] link to an issue
- [ ] add milestone (the current sprint)

### CODE REVIEW

**Review by mrBliss — COMMENTED** [review]
I'm curious why all that CPP is really needed. Both `ouroboros-network` and `cardano-ledger-specs` didn't need CPP at all.

Have you tried simply removing the imports that GHC 8.10.2 says are redundant? GHC became better in detecting redundant imports. I bet that you can remove them all and that it will still compile with GHC 8.6.5.

**Comment by CodiePP** [comment]
> I'm curious why all that CPP is really needed. Both `ouroboros-network` and `cardano-ledger-specs` didn't need CPP at all.
> 
> Have you tried simply removing the imports that GHC 8.10.2 says are redundant? GHC became better in detecting redundant imports. I bet that you can remove them all and that it will still compile with GHC 8.6.5.

they are needed for GHC 8.6.5. And, the preprocessor statements make the changes very explicit. We can then drop them once we crossed over to 8.10.2.

**Review by mrBliss — DISMISSED** [review]
I can't compile this PR with GHC 8.10.2 using `cabal`. Since we use `cabal` in `cardano-node`, we must make sure that this repo compiles with `cabal`.

* There is a dependency on `Win32-network`, which has a `base >= 4.5  && < 4.13` constraint and GHC-8.10.2 comes with `base-4.14.1.0`. You'll have to wait until https://github.com/input-output-hk/ouroboros-network/pull/2539 is merged. I see `allower-newer` is set to `true` in the `stack.yaml` file. I would recommend against this, because this is just hiding dependency conflicts. So when you want to propagate this PR to `cardano-node`, you'll actually have to deal with all these conflicts, since `cardano-node` uses `cabal`.

* There are similar constraints on `base` within this repository that prevent it from compiling with GHC-8.10.2. See `lobemo-examples.cabal` and `lobemo-backend-ekg.cabal`.

* There is a dependency on [`libsystemd-journal`](https://hackage.haskell.org/package/libsystemd-journal), but no version of that package exists that is compatible with GHC-8.10.2: it has a `base >=4.6 && <4.13` constraint.

* Just like I said, *no CPP is needed* to be compatible with GHC 8.10.2.
  > they are needed for GHC 8.6.5.

  You say they are "needed", but you have not tried what I said! I have tried removing all imports that GHC 8.10.2 marked as redundant, and it *still compiles with GHC 8.6.5*. As I said, GHC became better in detecting redundant imports. That's why they weren't previously marked as redundant, but they already were redundant. I can open a PR with my changes if you don't believe me.

**Comment by CodiePP** [comment]
cherry-picked Thomas' commits and added further cleanups.
let's have it checked by CI.

File: examples/complex/Main.lhs [hunk 1]
```diff
@@ -31,9 +31,13 @@ import qualified Data.HashMap.Strict as HM
 import           Data.Maybe (isJust)
 import           Data.Text (Text, pack)
 #ifdef ENABLE_OBSERVABLES
+#ifdef RUN_ProcObseverSTM
 import           Control.Monad (forM)
 import           GHC.Conc.Sync (atomically, STM, TVar, newTVar, readTVar, writeTVar)
+#endif
+#endif
```
Comment (mrBliss) [inline examples/complex/Main.lhs:38]: I'm not familiar with the details, but wouldn't it be possible to drop the outer `ifdef`?

File: iohk-monitoring/test/Cardano/BM/Test/Configuration.lhs [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 %if style == newcode
 \begin{code}
-{-# LANGUAGE CPP               #-}
+{- # LANGUAGE CPP               # -}
```
Comment (mrBliss) [inline iohk-monitoring/test/Cardano/BM/Test/Configuration.lhs:7]: I would just delete it

**Review by mrBliss — APPROVED** [review]
LGTM

File: stack.yaml [hunk 3]
```diff
@@ -32,13 +32,16 @@ extra-deps:
...
-
   - git: https://github.com/input-output-hk/ouroboros-network
     commit: 3b92f6e936e004753b46f3efabfbbb8539163ea1
     subdirs:
         - Win32-network
 
+  - git: https://github.com/CodiePP/libsystemd-journal
+    commit: 4371464d5ae3fe1df4e3e77ab33156ad2d3bf1bd
+    subdirs:
+        - .
```
Comment (mrBliss) [inline stack.yaml:43]: You can omit these two lines if you want.

**Comment by CodiePP** [comment]
there will be a change to `libsystemd` for 8.10.2 compatibility: https://github.com/ocharles/libsystemd-journal/pull/21
let's wait for this one.

File: cabal.project [hunk 4]
```diff
@@ -20,15 +20,21 @@ package iohk-monitoring
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/ouroboros-network
-  tag: 3b92f6e936e004753b46f3efabfbbb8539163ea1
-  --sha256: 1hg4xs9aj40xgkm9y8b601x0n4sndj71ppzym34gfq5jgs950lf1
+  tag: 5f79c1f2699d4af41fcdcb7d054d4c87573dd675
+  --sha256: 0ar09rmkkkkz90wfwhdnhfisl8ipfndn39adcmmjm7dambf851g0
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply cabal.project:24]: FYI, that PR was just merged, you can use https://github.com/input-output-hk/ouroboros-network/commit/64351aaa3b27fd5a28fc7d1f9c48f9e5a815dab2 as the tag for the dependency.

**Comment by deepfire** [comment]
So, https://github.com/input-output-hk/ouroboros-network/pull/2539 was merged, so this can be merged as well I presume, @mrBliss, @CodiePP ?

**Review by deepfire — APPROVED** [review]
LGTM

**Pull request merged**